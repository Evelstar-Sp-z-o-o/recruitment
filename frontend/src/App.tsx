import { FC } from 'react';

import Root from '@/src/pages/Root';
import AppProviders from '@/src/utils/providers/AppProviders';

import './styles/main.scss';

const App: FC = () => {
  return (
    <AppProviders>
      <Root />
    </AppProviders>
  );
};

export default App;
