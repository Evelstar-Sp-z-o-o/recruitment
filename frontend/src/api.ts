import type { Post } from './types';

export const BASE_API_URL = 'http://localhost:3000/api';

const getPosts = async () => {
  const ENDPOINT = '/posts';

  const res = await fetch(BASE_API_URL + ENDPOINT);

  if (!res.ok) {
    throw new Error('Something went wrong when loading Posts list.');
  }

  const data = await res.json();

  return data as unknown as Post[];
};

const getSinglePost = async (postId: string) => {
  const ENDPOINT = `/posts/${postId}`;

  const res = await fetch(BASE_API_URL + ENDPOINT);

  if (!res.ok) {
    throw new Error('Something went wrong when loading Post details.');
  }

  const data = await res.json();
  return data as unknown as Post;
};

const deletePost = async (postId: string) => {
  const ENDPOINT = `/posts/${postId}`;

  const res = await fetch(BASE_API_URL + ENDPOINT, {
    method: 'DELETE',
  });

  if (!res.ok) {
    throw new Error('Something went wrong when deleting post.');
  }

  const data = await res.json();
  return data as unknown as Post;
};

const updatePost = async (postId: string, postData: Partial<Post>) => {
  const ENDPOINT = `/posts/${postId}`;

  const res = await fetch(BASE_API_URL + ENDPOINT, {
    method: 'PATCH',
    headers: {
      'Content-Type': 'application/json',
    },
    body: JSON.stringify(postData),
  });

  if (!res.ok) {
    throw new Error('Something went wrong when updating post.');
  }

  const data = await res.json();
  return data as unknown as Post;
};

const createPost = async (postData: Partial<Post>) => {
  const ENDPOINT = `/posts`;

  const res = await fetch(BASE_API_URL + ENDPOINT, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    body: JSON.stringify(postData),
  });

  if (!res.ok) {
    throw new Error('Something went wrong when creating post.');
  }

  const data = await res.json();

  return data as unknown as Post;
};

export { getPosts, getSinglePost, updatePost, createPost, deletePost };
