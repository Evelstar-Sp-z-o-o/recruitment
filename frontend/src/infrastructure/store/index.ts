import { useDispatch, useSelector } from 'react-redux';

import { configureStore } from '@reduxjs/toolkit';

import apiService from './services/apiService';
import userSlice from './slices/user';

const store = configureStore({
  reducer: {
    [apiService.reducerPath]: apiService.reducer,
    user: userSlice.reducer,
  },
  middleware: (getDefaultMiddleware) => getDefaultMiddleware().concat(apiService.middleware),
});

export type RootState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch;

export const useAppDispatch: () => AppDispatch = useDispatch;
export const useAppSelector = useSelector.withTypes<RootState>();

export default store;
