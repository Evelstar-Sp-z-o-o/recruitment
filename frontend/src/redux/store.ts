import { configureStore } from '@reduxjs/toolkit';

import responseModalReducer from './responseModalSlice';

const store = configureStore({
  reducer: {
    responseModal: responseModalReducer,
  },
});

export type RootState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch;

export default store;
