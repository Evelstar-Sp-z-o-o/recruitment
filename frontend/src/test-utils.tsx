import { ReactNode } from 'react';

import AppProviders from '@/src/utils/providers/AppProviders';
import { render } from '@testing-library/react';

const AllTheProviders = ({ children }: { children: ReactNode }) => {
  return <AppProviders>{children}</AppProviders>;
};

const customRender = (ui, options?) => render(ui, { wrapper: AllTheProviders, ...options });

export * from '@testing-library/react';

export { customRender as render };
