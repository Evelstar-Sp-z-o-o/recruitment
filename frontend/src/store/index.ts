import { postsApi } from '@/src/store/api/posts';
import { configureStore, createSlice } from '@reduxjs/toolkit';

const initialUser: null | string = null;
const initialLogin = false;

const userSlice = createSlice({
  name: 'user',
  initialState: initialUser,
  reducers: {
    setUser(_state, action) {
      return action.payload;
    },
  },
});

const loginSlice = createSlice({
  name: 'login',
  initialState: initialLogin,
  reducers: {
    setLogin(_state, action) {
      return action.payload;
    },
  },
});

export const { setUser } = userSlice.actions;
export const { setLogin } = loginSlice.actions;

export * from './api/posts.ts';

export const store = configureStore({
  reducer: {
    [postsApi.reducerPath]: postsApi.reducer,
    user: userSlice.reducer,
    login: loginSlice.reducer,
  },
  middleware: (getDefaultMiddleware) => getDefaultMiddleware().concat(postsApi.middleware),
});

export type RootState = ReturnType<typeof store.getState>;
