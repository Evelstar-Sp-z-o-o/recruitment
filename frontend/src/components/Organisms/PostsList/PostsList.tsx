import { FC, Fragment, useEffect, useState } from 'react';
import { useTranslation } from 'react-i18next';
import { useSelector } from 'react-redux';

import PostListAlerts from '@/src/components/Molecules/PostListAlerts/PostListAlerts';
import { IPost, useRemovePostMutation, RootState } from '@/src/store';
import { formatDate } from '@/src/utils/helpers/formatDate';
import DeleteForeverIcon from '@mui/icons-material/DeleteForever';
import EditIcon from '@mui/icons-material/Edit';
import TextsmsOutlinedIcon from '@mui/icons-material/TextsmsOutlined';
import Avatar from '@mui/material/Avatar';
import Box from '@mui/material/Box';
import Card from '@mui/material/Card';
import CardActions from '@mui/material/CardActions';
import CardContent from '@mui/material/CardContent';
import CardHeader from '@mui/material/CardHeader';
import Divider from '@mui/material/Divider';
import IconButton from '@mui/material/IconButton';
import Typography from '@mui/material/Typography';
import { blue } from '@mui/material/colors';

interface IPostsListProps {
  posts: IPost[];
  handleEdit: () => void;
}

const PostsList: FC<IPostsListProps> = ({ posts, handleEdit }) => {
  const user = useSelector<RootState>((state) => state.user);
  const [removePost, { isError, isSuccess }] = useRemovePostMutation();
  const { t } = useTranslation();
  const [sortedPosts, setSortedPosts] = useState<IPost[]>([]);
  const [open, setOpen] = useState<boolean>(false);
  const [postId, setPostId] = useState<null | number>(null);
  const [alert, setAlert] = useState<boolean>(false);
  const [authorAlert, setAuthorAlert] = useState<boolean>(false);

  useEffect(() => {
    if (isError || isSuccess) {
      setAlert(true);
    }
  }, [isError, isSuccess]);

  const handleCloseSnackbar = () => {
    setAlert(false);
  };

  const handleClose = () => {
    setOpen(false);
  };

  const handleAlertClose = () => {
    setAuthorAlert(false);
    setOpen(false);
  };

  const handleDelete = (post: IPost) => {
    if (post.data.author !== user) {
      setPostId(null);
      setAuthorAlert(true);

      return;
    }

    setOpen(true);
    setPostId(post.id);
  };

  const handleConfirmDelete = () => {
    setOpen(false);
    removePost(postId);
    setPostId(null);
  };

  useEffect(() => {
    if (posts) {
      const sorted = [...posts];
      sorted.sort((a, b) => {
        return b.data.created - a.data.created;
      });

      setSortedPosts([...sorted]);
    }
  }, [posts]);

  return (
    <>
      {sortedPosts && sortedPosts.length > 0 ? (
        sortedPosts.map((post, index) => (
          <Fragment key={post.id}>
            <Card sx={{ p: '1.5rem 1rem' }}>
              <CardHeader
                avatar={
                  <Avatar sx={{ bgcolor: blue[500] }} alt="">
                    <TextsmsOutlinedIcon />
                  </Avatar>
                }
                title={post.data.author}
                subheader={`${formatDate(post.data.created)}${
                  post.data.created !== post.data.edited ? ' (edited)' : ''
                }`}
              />
              <CardContent>
                <Typography color="text.secondary">{post.data.body}</Typography>
              </CardContent>
              <CardActions disableSpacing sx={{ justifyContent: 'flex-end' }}>
                <IconButton
                  disabled={!user || user?.toLowerCase() !== post.data.author.toLowerCase()}
                  onClick={() => handleEdit(post)}
                >
                  <EditIcon />
                </IconButton>
                <IconButton
                  disabled={!user || user?.toLowerCase() !== post.data.author.toLowerCase()}
                  onClick={() => handleDelete(post)}
                >
                  <DeleteForeverIcon />
                </IconButton>
              </CardActions>
            </Card>
            {index !== posts.length - 1 ? <Divider /> : null}
          </Fragment>
        ))
      ) : (
        <Box maxWidth="dm" className="center">
          {t('noPosts')}
        </Box>
      )}
      <PostListAlerts
        open={open}
        handleClose={handleClose}
        handleAlertClose={handleAlertClose}
        authorAlert={authorAlert}
        alert={alert}
        handleCloseSnackbar={handleCloseSnackbar}
        handleConfirmDelete={handleConfirmDelete}
        isSuccess={isSuccess}
      />
    </>
  );
};

export default PostsList;
