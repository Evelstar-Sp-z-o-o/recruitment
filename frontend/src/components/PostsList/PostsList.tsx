import { useEffect } from 'react';
import { useSelector, useDispatch } from 'react-redux';

import Post from '@/src/components/Post';
import { RootState, AppDispatch } from '@/src/store/postStore';
import { fetchPosts } from '@/src/store/postsSlice';
import { Box, Typography } from '@mui/material';

const PostsList = () => {
  const dispatch = useDispatch<AppDispatch>();
  const postsState = useSelector((state: RootState) => state.posts);
  const { posts, status, error } = postsState;

  useEffect(() => {
    if (status === 'idle') {
      dispatch(fetchPosts());
    }
  }, [status, dispatch]);

  return (
    <Box
      sx={{
        display: 'flex',
        flexDirection: 'column',
        justifyContent: 'center',
        alignItems: 'center',
        margin: '8px 0',
      }}
    >
      {status === 'loading' && <Typography>Loading...</Typography>}
      {status === 'failed' && <Typography>{error}</Typography>}
      {status === 'succeeded' && posts.length > 0 && (
        <Box
          sx={{
            display: 'flex',
            width: '100%',
            justifyContent: 'center',
            flexDirection: 'column',
            alignItems: 'center',
            mt: 2,
            gap: '18px',
            mb: 4,
          }}
        >
          {posts.map(({ id, data }) => (
            <Post key={id} author={data.author} text={data.body} postId={id} created={data.created} />
          ))}
        </Box>
      )}
    </Box>
  );
};

export default PostsList;
