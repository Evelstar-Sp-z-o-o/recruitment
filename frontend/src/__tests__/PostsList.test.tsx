import React from 'react';

import PostsList from '@/src/components/Organisms/PostsList/PostsList';
import { store, setUser } from '@/src/store';
import { IPost } from '@/src/store';
import { render, screen, fireEvent, waitFor } from '@/src/test-utils';
import '@testing-library/jest-dom';

const mockPosts: IPost[] = [
  {
    id: 1,
    data: {
      author: 'test@example.com',
      body: 'Test post content 1',
      created: Date.now(),
      edited: Date.now(),
    },
  },
  {
    id: 2,
    data: {
      author: 'another@example.com',
      body: 'Test post content 2',
      created: Date.now(),
      edited: Date.now(),
    },
  },
];

describe('PostsList', () => {
  vi.mock('react-i18next', async (importOriginal) => {
    const actual = await importOriginal();
    return {
      ...actual,
      useTranslation: () => ({
        t: (key: string) => key,
      }),
      I18nextProvider: ({ children }) => <div>{children}</div>,
    };
  });

  beforeEach(() => {
    store.dispatch(setUser('test@example.com'));
  });

  it('renders posts when there are posts', () => {
    render(<PostsList posts={mockPosts} handleEdit={vi.fn()} />);

    expect(screen.getByText('Test post content 1')).toBeInTheDocument();
    expect(screen.getByText('Test post content 2')).toBeInTheDocument();
  });

  it('renders no posts message when there are no posts', () => {
    render(<PostsList posts={[]} handleEdit={vi.fn()} />);

    expect(screen.getByText('noPosts')).toBeInTheDocument();
  });

  it('calls handleEdit when the edit button is clicked', () => {
    const handleEdit = vi.fn();

    render(<PostsList posts={mockPosts} handleEdit={handleEdit} />);

    const editButtons = screen.getAllByTestId('EditIcon');

    expect(editButtons[0]).not.toBeDisabled();

    fireEvent.click(editButtons[0]);

    expect(handleEdit).toHaveBeenCalledWith(mockPosts[0]);
  });

  it('calls removePost mutation when confirm delete', async () => {
    render(<PostsList posts={mockPosts} handleEdit={vi.fn()} />);

    const deleteButton = screen.getAllByTestId('DeleteForeverIcon')[0];
    fireEvent.click(deleteButton);

    const confirmButton = screen.getByRole('button', { name: /delete.confirm.button.confirm/i });
    fireEvent.click(confirmButton);

    await waitFor(() => {
      expect(screen.getByText(/delete.alert/i)).toBeInTheDocument();
    });
  });

  it('should match snapshot', () => {
    const { asFragment } = render(<PostsList posts={mockPosts} handleEdit={vi.fn()} />);
    expect(asFragment()).toMatchSnapshot();
  });
});
