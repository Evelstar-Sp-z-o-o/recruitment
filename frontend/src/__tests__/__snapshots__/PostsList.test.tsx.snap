// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PostsList > should match snapshot 1`] = `
<DocumentFragment>
  <div>
    <div
      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-1x3n8rb-MuiPaper-root-MuiCard-root"
    >
      <div
        class="MuiCardHeader-root css-185gdzj-MuiCardHeader-root"
      >
        <div
          class="MuiCardHeader-avatar css-1ssile9-MuiCardHeader-avatar"
        >
          <div
            class="MuiAvatar-root MuiAvatar-circular MuiAvatar-colorDefault css-4vb5ob-MuiAvatar-root"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="TextsmsOutlinedIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M20 2H4c-1.1 0-2 .9-2 2v18l4-4h14c1.1 0 2-.9 2-2V4c0-1.1-.9-2-2-2zm0 14H6l-2 2V4h16v12zM7 9h2v2H7zm4 0h2v2h-2zm4 0h2v2h-2z"
              />
            </svg>
          </div>
        </div>
        <div
          class="MuiCardHeader-content css-1qbkelo-MuiCardHeader-content"
        >
          <span
            class="MuiTypography-root MuiTypography-body2 MuiCardHeader-title css-et1ao3-MuiTypography-root"
          >
            test@example.com
          </span>
          <span
            class="MuiTypography-root MuiTypography-body2 MuiCardHeader-subheader css-83ijpv-MuiTypography-root"
          >
            26.07.2024, 23:04
          </span>
        </div>
      </div>
      <div
        class="MuiCardContent-root css-46bh2p-MuiCardContent-root"
      >
        <p
          class="MuiTypography-root MuiTypography-body1 css-1pnmrwp-MuiTypography-root"
        >
          Test post content 1
        </p>
      </div>
      <div
        class="MuiCardActions-root css-5rkuwp-MuiCardActions-root"
      >
        <button
          class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
          tabindex="0"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="EditIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75L3 17.25zM20.71 7.04c.39-.39.39-1.02 0-1.41l-2.34-2.34a.9959.9959 0 0 0-1.41 0l-1.83 1.83 3.75 3.75 1.83-1.83z"
            />
          </svg>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
        <button
          class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
          tabindex="0"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="DeleteForeverIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zm2.46-7.12 1.41-1.41L12 12.59l2.12-2.12 1.41 1.41L13.41 14l2.12 2.12-1.41 1.41L12 15.41l-2.12 2.12-1.41-1.41L10.59 14l-2.13-2.12zM15.5 4l-1-1h-5l-1 1H5v2h14V4z"
            />
          </svg>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
    <hr
      class="MuiDivider-root MuiDivider-fullWidth css-9mgopn-MuiDivider-root"
    />
    <div
      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-1x3n8rb-MuiPaper-root-MuiCard-root"
    >
      <div
        class="MuiCardHeader-root css-185gdzj-MuiCardHeader-root"
      >
        <div
          class="MuiCardHeader-avatar css-1ssile9-MuiCardHeader-avatar"
        >
          <div
            class="MuiAvatar-root MuiAvatar-circular MuiAvatar-colorDefault css-4vb5ob-MuiAvatar-root"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="TextsmsOutlinedIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M20 2H4c-1.1 0-2 .9-2 2v18l4-4h14c1.1 0 2-.9 2-2V4c0-1.1-.9-2-2-2zm0 14H6l-2 2V4h16v12zM7 9h2v2H7zm4 0h2v2h-2zm4 0h2v2h-2z"
              />
            </svg>
          </div>
        </div>
        <div
          class="MuiCardHeader-content css-1qbkelo-MuiCardHeader-content"
        >
          <span
            class="MuiTypography-root MuiTypography-body2 MuiCardHeader-title css-et1ao3-MuiTypography-root"
          >
            another@example.com
          </span>
          <span
            class="MuiTypography-root MuiTypography-body2 MuiCardHeader-subheader css-83ijpv-MuiTypography-root"
          >
            26.07.2024, 23:04
          </span>
        </div>
      </div>
      <div
        class="MuiCardContent-root css-46bh2p-MuiCardContent-root"
      >
        <p
          class="MuiTypography-root MuiTypography-body1 css-1pnmrwp-MuiTypography-root"
        >
          Test post content 2
        </p>
      </div>
      <div
        class="MuiCardActions-root css-5rkuwp-MuiCardActions-root"
      >
        <button
          class="MuiButtonBase-root Mui-disabled MuiIconButton-root Mui-disabled MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
          disabled=""
          tabindex="-1"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="EditIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75L3 17.25zM20.71 7.04c.39-.39.39-1.02 0-1.41l-2.34-2.34a.9959.9959 0 0 0-1.41 0l-1.83 1.83 3.75 3.75 1.83-1.83z"
            />
          </svg>
        </button>
        <button
          class="MuiButtonBase-root Mui-disabled MuiIconButton-root Mui-disabled MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
          disabled=""
          tabindex="-1"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="DeleteForeverIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zm2.46-7.12 1.41-1.41L12 12.59l2.12-2.12 1.41 1.41L13.41 14l2.12 2.12-1.41 1.41L12 15.41l-2.12 2.12-1.41-1.41L10.59 14l-2.13-2.12zM15.5 4l-1-1h-5l-1 1H5v2h14V4z"
            />
          </svg>
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
